File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/C Programs/1_insert_operations.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node { int data; struct Node* next; } Node;
Node* cn(int d) { Node* n = malloc(sizeof(Node)); n->data = d; n->next = NULL; return n; }
Node* iaf(Node* h, int d) { Node* n = cn(d); n->next = h; return n; }
Node* ial(Node* h, int d) { Node* n = cn(d); if (!h) return n; Node* t = h; while (t->next) t = t->next; t->next = n; return h; }
Node* iaa(Node* h, int d, int p) { if (p == 1) return iaf(h, d); Node* n = cn(d), *t = h; for (int i = 1; i < p-1 && t; i++) t = t->next; if (!t) return ial(h, d); n->next = t->next; t->next = n; return h; }
void disp(Node* h) { while (h) printf("%d -> ", h->data), h = h->next; printf("NULL\n"); }
int main() {
    Node* h = NULL; int c, d, p;
    while (scanf("%d", &c) && c != 6) {
        if (c == 1) scanf("%d", &d), h = iaf(h, d);
        else if (c == 2) scanf("%d", &d), h = ial(h, d);
        else if (c == 3) scanf("%d%d", &d, &p), h = iaa(h, d, p);
        else if (c == 4) disp(h);
    }
    return 0;
}

File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/Python Programs/1_insert_operations.py
class Node:
    def __init__(self, d): self.data, self.next = d, None

class LL:
    def __init__(self): self.head = None
    def iaf(self, d): n = Node(d); n.next = self.head; self.head = n
    def ial(self, d): 
        n = Node(d)
        if not self.head: self.head = n
        else: t = self.head; [t := t.next for _ in iter(lambda: t.next, None)]; t.next = n
    def iaa(self, d, p):
        if p == 1: return self.iaf(d)
        t = self.head; [t := t.next for _ in range(p-2) if t]; 
        if not t: return self.ial(d)
        n = Node(d); n.next = t.next; t.next = n
    def disp(self): 
        if not self.head: print('Empty'); return
        t = self.head; print(' -> '.join([str(t.data)] + [str((t := t.next).data) for _ in iter(lambda: t.next, None)]) + ' -> None')

ll = LL()
while True:
    c = input("1:First 2:Last 3:Any 4:Show 5:Exit: ")
    if c == '1': ll.iaf(int(input()))
    elif c == '2': ll.ial(int(input()))
    elif c == '3': ll.iaa(int(input()), int(input()))
    elif c == '4': ll.disp()
    elif c == '5': break